osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs

```diff
@@ -102,19 +110,84 @@ public override double DifficultyValue()
             double difficulty = 0;
             double weight = 1;
 
-            // Sections with 0 strain are excluded to avoid worst-case time complexity of the following sort (e.g. /b/2351871).
-            // These sections will not contribute to the difficulty.
-            var peaks = GetCurrentStrainPeaks().Where(p => p > 0);
-
             // Difficulty is the weighted sum of the highest strains from every section.
             // We're sorting from highest to lowest strain.
-            foreach (double strain in peaks.OrderDescending())
+            foreach (double strain in GetCurrentStrainsSorted())
             {
                 difficulty += strain * weight;
                 weight *= DecayWeight;
             }
 
             return difficulty;
         }
+
+        protected List<double> GetCurrentStrainsSorted()
+        {
+            List<double> strains;
+
+            // If no saved strains - calculate them from 0, and save them after that
+            if (savedSortedStrains == null || savedSortedStrains.Count == 0)
```

Comment:
The zero-count part of this check is super confusing to me. What scenario is this supposed to handle..?

Reply:
This function is called whenever `DifficultyValue` is called
So this covers pretty much every scenario except progressive calculations, preserving the same behavior as before (just straight up sort all the strains)